### ISSUE TITLE
Around filters

### ISSUE THREAD

**Issue by kpsrinivas** [bigcommerce/gruf#89]
## What? Why?
Can we get hooks like rails controller in this ? Interceptors are good but hooks can help set instance variables kind of thing. 
## Other Notables
Git this from a gist. But the idea is to do something like this:

```
module Util
  module ExecutionHooks
  # this method is invoked whenever a new instance method is added to a class
    def method_added(method_name)
      return if hooks.include?(method_name) || after_hooks.include?(method_name) || hooked_methods.include?(method_name)
      add_hooks_to(method_name)
    end

    def before_filter(method_name)
      hooks << method_name
    end

    def after_filter(method_name)
      after_hooks << method_name
    end

    def hooks
      @hooks ||= []
    end
    
    def after_hooks
      @after_hooks ||=[]
    end

    private

    def hooked_methods
      @hooked_methods ||= []
    end

    def add_hooks_to(method_name)

      hooked_methods << method_name

      original_method = instance_method(method_name)

      define_method(method_name) do |*args, &block|

        self.class.hooks.each do |hook|
          method(hook).call
        end

        response = original_method.bind(self).call(*args, &block)

        self.class.after_hooks.each do |hook|
          method(hook).call
        end

        response
        
      end
    end
  end
end
```

**Comment by timdorr** [comment]
You could use [ActiveSupport::Callbacks](https://api.rubyonrails.org/classes/ActiveSupport/Callbacks/ClassMethods.html) for this, since AS is already a dependency.

**Comment by timdorr** [comment]
OK, got it working in userland. I'll package this up into a PR sometime soon. For now, I've got a few things in place:

Here's the callback module with the comments stripped and a small change to `define_callback`, since Gruf controllers don't have a `processed?` method.
```ruby
# lib/gruf/controllers/callbacks.rb
module Gruf
  module Controllers
    module Callbacks
      extend ActiveSupport::Concern

      include ActiveSupport::Callbacks

      included do
        define_callbacks :process_action
      end

      def process_action(*)
        run_callbacks(:process_action) do
          super
        end
      end

      module ClassMethods
        def _normalize_callback_options(options)
          _normalize_callback_option(options, :only, :if)
          _normalize_callback_option(options, :except, :unless)
        end

        def _normalize_callback_option(options, from, to) # :nodoc:
          if from = options[from]
            _from = Array(from).map(&:to_s).to_set
            from = proc { |c| _from.include? c.action_name }
            options[to] = Array(options[to]).unshift(from)
          end
        end

        def _insert_callbacks(callbacks, block = nil)
          options = callbacks.extract_options!
          _normalize_callback_options(options)
          callbacks.push(block) if block
          callbacks.each do |callback|
            yield callback, options
          end
        end

        [:before, :after, :around].each do |callback|
          define_method "#{callback}_action" do |*names, &blk|
            _insert_callbacks(names, blk) do |name, options|
              set_callback(:process_action, callback, name, options)
            end
          end

          define_method "prepend_#{callback}_action" do |*names, &blk|
            _insert_callbacks(names, blk) do |name, options|
              set_callback(:process_action, callback, name, options.merge(prepend: true))
            end
          end

          define_method "skip_#{callback}_action" do |*names|
            _insert_callbacks(names) do |name, options|
              skip_callback(:process_action, callback, name, options)
            end
          end

          alias_method :"append_#{callback}_action", :"#{callback}_action"
        end
      end
    end
  end
end
```

Then I've set up an "ApplicationRpc" controller (I don't use the "Controller" postfix so they don't overlap with my Rails controllers) to inherit from:

```ruby
# app/rpc/application_rpc.rb
class ApplicationRpc < ::Gruf::Controllers::Base
  include Gruf::Controllers::Callbacks

  before_action :authorize_request!

  def authorize_request!
    fail!(:unauthorized, :bad_auth, 'go login, plz')
  end
end
```

And here's a typical controller:

```ruby
# app/rpc/auth_rpc.rb
class AuthRpc < ApplicationRpc
  bind ::Foo::Users::Service

  def current_user
    Foo::User.new(name: 'Test', email: 'test@test.test')
  end
end
```

Works like a treat!

**Comment by kpsrinivas** [comment]
Thanks for this. I am using this in a standalone ruby application with Rom datamapper. Do you recommend me to install active_support gem ? But process looks clean.

**Comment by kpsrinivas** [comment]
Just saw that gruf already has activesupport in gemspec. Thanks for this.

**Comment by splittingred** [comment]
So filters in Rails are pretty controversial. There's a strong argument that they violate [SOLID principles](https://en.wikipedia.org/wiki/SOLID) of software design, encouraging engineers to write logic in controllers. Controllers IMO should be relegated to handling protocol responses and delegating serialization to serializers. I'd rather not have functionality like this in gruf core that encourages that sort of behavior. 

It's also important to note that gruf _rebuilds_ `Gruf::Controller` instances for every request. The reason it does this is that gRPC controllers are not thread-safe; gRPC's core library instantiates "Service" classes at proc start, and then manages a thread pool to delegate off to them.

To ensure that there's no ambiguity and potential thread-safe instance vars on classes, Gruf binds methods to the service classes that create new controllers from requests every time. (You can see this in the `ServiceBinder` [class here](https://github.com/bigcommerce/gruf/blob/7038ae67574c4240651f379207c2c22c9b34f0b2/lib/gruf/controllers/service_binder.rb#L56). This also allows us to use instance variables (such as what the error handling methods use) in those controllers.

TL;DR: instance variables in controllers don't persist across requests. Therefore, allowing filters would be for a per-request basis only. That means their only benefit is to DRY up code in multiple methods; however, a much more preferable approach here is the interceptor pattern, which properly establishes FIFO ordering, promotes SRP (single-responsibility principles), and ensures consistency of operations.

All that said, feel free to create a gem that does this. I'd just rather not have it in gruf-core, if that's okay. Thanks!
